rescript/tests/test_get_unite.py

```diff
@@ -0,0 +1,105 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2019-2023, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import pkg_resources
+import tempfile
+import pandas.core.frame
+import q2_types.feature_data
+from qiime2.plugin.testing import TestPluginBase
+from rescript.get_unite import (
+    UNITE_DOIS,
+    _unite_get_url,
+    _unite_get_tgz,
+    _unite_get_artifacts,
+    get_unite_data,
+)
+
+from urllib.request import urlopen
+from unittest.mock import patch
+import os
+
+
+class TestGetUNITE(TestPluginBase):
+    package = "rescript.tests"
+
+    def setUp(self):
+        super().setUp()
+        self.unitefile = pkg_resources.resource_filename(
+            "rescript.tests", "data/unitefile.tgz"
+        )
+        self.unitefile_no_dev = pkg_resources.resource_filename(
+            "rescript.tests", "data/unitefile_no_dev.tgz"
+        )
+
+    # Requires internet access
+    def test_unite_get_url(self):
+        # for all combinations...
+        for v in UNITE_DOIS.keys():
+            for tg in UNITE_DOIS[v].keys():
+                for s in UNITE_DOIS[v][tg].keys():
+                    # ... try to get the URL
+                    url = _unite_get_url(v, tg, s)
+                    urlopen(url)
+        self.assertTrue(True)
+
+    def test_unite_get_tgz(self):
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            # Simulate a successful download
+            with patch("rescript.get_unite.requests.get") as mock_requests_get:
```

Comment:
Sorry if I am missing the obvious, but I really do not understand what is happening within this block of code.

Reply:
Yeah, I don't either. I'm way out of my depth here, but I'll try to explain:

```python
try:
    response = requests.get(url, stream=True)       # The main response object needs to be mocked
    # Save .tgz file
    unite_file_path = os.path.join(download_path, "unitefile.tar.gz")
    with open(unite_file_path, "wb") as f:
        for chunk in response.iter_content(chunk_size=8192): # response.iter_content needs to be mocked
            if chunk:
                f.write(chunk)
                file_size += len(chunk)
    # Check if the downloaded size matches the expected size
    if file_size == int(response.headers.get("content-length", 0)): # response.headers.get needs to be mocked
        return unite_file_path  # done!
```

So, fake `request.get()` to return to fake `response`
But that has to return fake content from `response.iter_content()`
That content has to match `response.headers.get("content-length", 0)`, which also needs to be mocked

Reply:
OK, this works, now with the addition of `Mock()`

```python
def test_unite_get_tgz2(self):
    with tempfile.TemporaryDirectory() as tmpdirname:
        # mock the response object
        mock_response = Mock()
        mock_response.iter_content.return_value = [b"mock"]
        mock_response.headers.get.return_value = "4"  # matches content
        # mock successful download
        with patch("requests.get", return_value=mock_response):
            _unite_get_tgz("fakeURL", tmpdirname)
        # real failed download
        with self.assertRaisesRegex(ValueError, "File incomplete on try"):
            _unite_get_tgz("https://files.plutof.ut.ee/nope", tmpdirname)
```

Changing `return_value = "3"` makes it fail, as content no longer matches it's header.

Am I on the right track?